File: src/GameCore/Graph/GraphLoader.cs
using System.Text.Json;
using GameCore.Tag;

namespace GameCore.Graph;

/// <summary>
/// GraphLoader — 三图加载器。解析 JSON → 创建节点 → 解析边条件 → 连接。
/// </summary>
public sealed class GraphLoader
{
    private readonly TagConditionParser _parser;
    private readonly TagRegistry _registry;

    public GraphLoader(TagConditionParser parser, TagRegistry registry)
    {
        _parser = parser;
        _registry = registry;
    }

    // ==================== 种族进化图 ====================

    public GraphEngine<RaceData> LoadRaceGraph(string jsonPath)
        => LoadRaceGraphFromText(File.ReadAllText(jsonPath));

    public GraphEngine<RaceData> LoadRaceGraphFromText(string json)
    {
        var engine = new GraphEngine<RaceData>(_parser);
        using var doc = JsonDocument.Parse(json);

        // 第一遍：创建全部节点
        foreach (var prop in doc.RootElement.EnumerateObject())
        {
            var node = prop.Value;
            var data = new RaceData(
                GetString(node, "name", prop.Name),
                ParseStringSet(node, "tagIds"),
                ParseFloatMap(node, "baseStats"),
                ParseStringSet(node, "conflictTags"));
            engine.AddNode(new RaceNode(prop.Name, data));
        }

        // 第二遍：连接边
        foreach (var prop in doc.RootElement.EnumerateObject())
            ConnectEdges(engine, prop.Name, prop.Value);

        return engine;
    }

    // ==================== 职业转职图 ====================

    public GraphEngine<ClassData> LoadClassGraph(string jsonPath)
        => LoadClassGraphFromText(File.ReadAllText(jsonPath));

    public GraphEngine<ClassData> LoadClassGraphFromText(string json)
    {
        var engine = new GraphEngine<ClassData>(_parser);
        using var doc = JsonDocument.Parse(json);

        foreach (var prop in doc.RootElement.EnumerateObject())
        {
            var node = prop.Value;
            var data = new ClassData(
                GetString(node, "name", prop.Name),
                ParseStringSet(node, "tagIds"),
                node.TryGetProperty("skillTreeRoot", out var st) ? st.GetString() : null,
                ParseFloatMap(node, "statGrowth"));
            engine.AddNode(new ClassNode(prop.Name, data));
        }

        foreach (var prop in doc.RootElement.EnumerateObject())
            ConnectEdges(engine, prop.Name, prop.Value);

        return engine;
    }

    // ==================== 任务图 ====================

    public GraphEngine<QuestData> LoadQuestGraph(string jsonPath)
        => LoadQuestGraphFromText(File.ReadAllText(jsonPath));

    public GraphEngine<QuestData> LoadQuestGraphFromText(string json)
    {
        var engine = new GraphEngine<QuestData>(_parser);
        using var doc = JsonDocument.Parse(json);

        foreach (var prop in doc.RootElement.EnumerateObject())
        {
            var node = prop.Value;
            ITagCondition accept = node.TryGetProperty("acceptCondition", out var ac) && ac.ValueKind == JsonValueKind.String
                ? _parser.Parse(ac.GetString())
                : null;
            var rewards = new Dictionary<string, int>();
            if (node.TryGetProperty("rewards", out var rw))
                foreach (var f in rw.EnumerateObject())
                    rewards[f.Name] = f.Value.ValueKind == JsonValueKind.Number ? f.Value.GetInt32() : 1;
            var data = new QuestData(
                GetString(node, "name", prop.Name),
                GetString(node, "description", ""),
                accept,
                node.TryGetProperty("minLevel", out var ml) ? ml.GetInt32() : 0,
                node.TryGetProperty("cooldown", out var cd) ? cd.GetInt32() : 0,
                ParseStringSet(node, "rewardTagIds"),
                rewards);
            engine.AddNode(new QuestNode(prop.Name, data));
        }

        foreach (var prop in doc.RootElement.EnumerateObject())
            ConnectEdges(engine, prop.Name, prop.Value);

        return engine;
    }

    // ==================== 公共 ====================

    private void ConnectEdges<T>(GraphEngine<T> engine, string fromId, JsonElement node)
    {
        if (!node.TryGetProperty("edges", out var edges)) return;
        foreach (var e in edges.EnumerateArray())
        {
            var to = e.GetProperty("to").GetString();
            ITagCondition condition = e.TryGetProperty("condition", out var c) && c.ValueKind == JsonValueKind.String
                ? _parser.Parse(c.GetString())
                : null;
            var requirements = new Dictionary<string, object>();
            if (e.TryGetProperty("requirements", out var req))
            {
                foreach (var f in req.EnumerateObject())
                    requirements[f.Name] = f.Value.ValueKind == JsonValueKind.Number ? (object)f.Value.GetInt32() : f.Value.GetString();
            }
            var bidirectional = e.TryGetProperty("bidirectional", out var bd) && bd.GetBoolean();
            var description = e.TryGetProperty("description", out var d) ? d.GetString() : null;
            engine.Connect(fromId, to, condition, requirements, bidirectional, description);
        }
    }

    private static string GetString(JsonElement node, string prop, string fallback)
        => node.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : fallback;

    private static HashSet<string> ParseStringSet(JsonElement node, string prop)
    {
        var set = new HashSet<string>();
        if (!node.TryGetProperty(prop, out var arr) || arr.ValueKind != JsonValueKind.Array) return set;
        foreach (var item in arr.EnumerateArray()) set.Add(item.GetString());
        return set;
    }

    private static Dictionary<string, float> ParseFloatMap(JsonElement node, string prop)
    {
        var map = new Dictionary<string, float>();
        if (!node.TryGetProperty(prop, out var obj) || obj.ValueKind != JsonValueKind.Object) return map;
        foreach (var f in obj.EnumerateObject())
            map[f.Name] = (float)f.Value.GetDouble();
        return map;
    }
}

File: canglan-backend/canglan-core/src/main/java/com/canglan/core/graph/GraphLoader.java
package com.canglan.core.graph;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import com.canglan.core.json.JsonReader;
import com.canglan.core.json.JsonValue;
import com.canglan.core.tag.TagCondition;
import com.canglan.core.tag.TagConditionParser;
import com.canglan.core.tag.TagRegistry;

/**
 * 三图加载器。解析 JSON → 创建节点 → 解析边条件 → 连接（两遍法）。
 * 对应 C# GraphLoader。
 */
public final class GraphLoader {

    private final TagConditionParser parser;
    private final TagRegistry registry;

    public GraphLoader(TagConditionParser parser, TagRegistry registry) {
        this.parser = parser;
        this.registry = registry;
    }

    public TagRegistry registry() { return registry; }

    // ==================== 种族进化图 ====================

    public GraphEngine<RaceData> loadRaceGraphFromText(String json) {
        GraphEngine<RaceData> engine = new GraphEngine<>(parser);
        JsonValue root = JsonReader.parse(json);

        // 第一遍：创建全部节点
        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            JsonValue node = entry.getValue();
            RaceData data = new RaceData(
                    node.getString("name", entry.getKey()),
                    parseStringSet(node, "tagIds"),
                    parseFloatMap(node, "baseStats"),
                    parseStringSet(node, "conflictTags"));
            engine.addNode(new RaceNode(entry.getKey(), data));
        }

        // 第二遍：连接边
        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            connectEdges(engine, entry.getKey(), entry.getValue());
        }
        return engine;
    }

    // ==================== 职业转职图 ====================

    public GraphEngine<ClassData> loadClassGraphFromText(String json) {
        GraphEngine<ClassData> engine = new GraphEngine<>(parser);
        JsonValue root = JsonReader.parse(json);

        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            JsonValue node = entry.getValue();
            ClassData data = new ClassData(
                    node.getString("name", entry.getKey()),
                    parseStringSet(node, "tagIds"),
                    node.getString("skillTreeRoot", null),
                    parseFloatMap(node, "statGrowth"));
            engine.addNode(new ClassNode(entry.getKey(), data));
        }

        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            connectEdges(engine, entry.getKey(), entry.getValue());
        }
        return engine;
    }

    // ==================== 任务图 ====================

    public GraphEngine<QuestData> loadQuestGraphFromText(String json) {
        GraphEngine<QuestData> engine = new GraphEngine<>(parser);
        JsonValue root = JsonReader.parse(json);

        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            JsonValue node = entry.getValue();
            JsonValue ac = node.get("acceptCondition");
            TagCondition accept = (ac != null && ac.isString()) ? parser.parse(ac.asString()) : null;
            Map<String, Integer> rewards = new HashMap<>();
            JsonValue rw = node.get("rewards");
            if (rw != null && rw.isObject()) {
                for (Map.Entry<String, JsonValue> f : rw.asObject().entrySet()) {
                    rewards.put(f.getKey(), f.getValue().isNumber() ? f.getValue().asInt() : 1);
                }
            }
            QuestData data = new QuestData(
                    node.getString("name", entry.getKey()),
                    node.getString("description", ""),
                    accept,
                    node.getInt("minLevel", 0),
                    node.getInt("cooldown", 0),
                    parseStringSet(node, "rewardTagIds"),
                    rewards);
            engine.addNode(new QuestNode(entry.getKey(), data));
        }

        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            connectEdges(engine, entry.getKey(), entry.getValue());
        }
        return engine;
    }

    // ==================== 公共 ====================

    private <T> void connectEdges(GraphEngine<T> engine, String fromId, JsonValue node) {
        JsonValue edges = node.get("edges");
        if (edges == null || !edges.isArray()) return;
        for (JsonValue e : edges.asArray()) {
            String to = e.getString("to", null);
            JsonValue c = e.get("condition");
            TagCondition condition = (c != null && c.isString()) ? parser.parse(c.asString()) : null;
            Map<String, Object> requirements = new HashMap<>();
            JsonValue req = e.get("requirements");
            if (req != null && req.isObject()) {
                for (Map.Entry<String, JsonValue> f : req.asObject().entrySet()) {
                    JsonValue v = f.getValue();
                    requirements.put(f.getKey(), v.isNumber() ? (Object) v.asInt() : v.asString());
                }
            }
            boolean bidirectional = e.getBoolean("bidirectional", false);
            String description = e.getString("description", null);
            engine.connect(fromId, to, condition, requirements, bidirectional, description);
        }
    }

    private static Set<String> parseStringSet(JsonValue node, String prop) {
        Set<String> set = new HashSet<>();
        JsonValue arr = node.get(prop);
        if (arr == null || !arr.isArray()) return set;
        for (JsonValue item : arr.asArray()) set.add(item.asString());
        return set;
    }

    private static Map<String, Double> parseFloatMap(JsonValue node, String prop) {
        Map<String, Double> map = new HashMap<>();
        JsonValue obj = node.get(prop);
        if (obj == null || !obj.isObject()) return map;
        for (Map.Entry<String, JsonValue> f : obj.asObject().entrySet()) {
            map.put(f.getKey(), f.getValue().asDouble());
        }
        return map;
    }
}
